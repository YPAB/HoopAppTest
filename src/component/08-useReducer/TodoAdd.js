import React from "react";
import { useForm } from "../../hooks/useForm";
import './styles.css'

export const TodoAdd = ({ handleAddTask }) => {

    const [{description},handleInputChange,reset ] = useForm({
        description:''

    });

    const handleSubmit = (e) =>{
        e.preventDefault();

        if (description.trim().length <= 1 ) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc:description,
            done: false
        };

        handleAddTask( newTodo );
        reset();
        
    }

    return (
        <>
             <h4> Agregar Tarea</h4>
                   <hr/>

                   <form onSubmit={ handleSubmit }>
                       <input
                        type='text'
                        name='description'
                        className='form-control'
                        placeholder='Aprender ...'
                        autoComplete='off'
                        value={ description }
                        onChange={ handleInputChange }
                       />

                       <button
                        type='submit'
                        className='btn btn-primary mt-1 btn-blck'
                       >
                        Agregar
                       </button>

                    </form>
        </>


    );

}