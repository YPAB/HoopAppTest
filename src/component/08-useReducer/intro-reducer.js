const tasks = [{
    id:1,
    todo:'Comprar Pan',
    done:false
}];

const todoReducer = ( state = tasks,action) => {

    if ( action?.type == 'create' ) {
        return [ ...state, action.payload ];
    }

    return state;
}

let todos = todoReducer();

const newTodo = {
    id:2,
    todo:'Comprar Noot',
    done:false
}


const agregarTodoAction = {
    type:'create',
    payload: newTodo
}

todos = todoReducer( todos, agregarTodoAction );

console.log( todos );