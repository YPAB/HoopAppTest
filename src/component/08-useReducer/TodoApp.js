import React,{useEffect, useReducer}from 'react';
import './styles.css'
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

export const TodoApp = () => {

    const init = () => {


        //si hay todos traelos o sino retorna un arreglo vacio
        return JSON.parse(localStorage.getItem('todos')) || [];


        /* return [{
            id: new Date().getTime(),
            desc:'Aprender React',
            done: false
        }]; */


    }

    /* const initialState = [ {
        id: new Date().getTime(),
        desc:'Aprender React',
        done: false
    }]; */

        //el dispath es una funcion que le mandamos una accion y ya va a saber a que reducer es el que hay que mandar.
    const [ todos,dispatch ] = useReducer( todoReducer,[], init );
    

    //console.log(todos);

    useEffect ( () => {
        localStorage.setItem('todos',JSON.stringify( todos ))

    },[todos]);

    const handleToggle = (todoId) => {

        const action = {
            type: 'toggle',
            payload: todoId
        }

        dispatch(action);

    }

    const handleAddTask = ( newTodo ) => {
        dispatch ( {
            type:'create',
            payload:newTodo
        });
    }


    const handleDelete = ( todoId ) => {
        const action = {
            type: 'delete',
            payload: todoId

        }

        dispatch(action);
    }
 
    return (
        <div>
            <h1>Tareas App: { todos.length }</h1>
            <hr/>

           <div className='row'>
               <div className='col-7'>
                   <TodoList 
                        todos = {todos} 
                        handleDelete= {handleDelete} 
                        handleToggle= { handleToggle} 
                    />
               </div>

               <div className='col-5'>
                  <TodoAdd
                    handleAddTask = { handleAddTask }
                  />

               </div>

           </div>

        </div>
    )
}