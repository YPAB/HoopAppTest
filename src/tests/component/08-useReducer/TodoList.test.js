import '@testing-library/jest-dom';
import { TodoList } from '../../../component/08-useReducer/TodoList'
import { shallow } from 'enzyme';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas en <TodoList></TodoList>', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(<TodoList 
        todos = {demoTodos}
        handleDelete = { handleDelete }
        handleToggle = { handleToggle }
    />)

    test('debe de mostrarse correctanmente', () => {

        expect( wrapper ).toMatchSnapshot();
    });

    test('debe de tener dos elementos', () => {

        expect( wrapper.find('TodoListItem').length ).toEqual(demoTodos.length);

        //Verificar que tengas sus props
        expect( wrapper.find('TodoListItem').at(0).prop('handleDelete') ).toEqual(expect.any(Function));


    })


})