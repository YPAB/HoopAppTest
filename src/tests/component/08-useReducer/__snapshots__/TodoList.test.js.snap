// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoList></TodoList> debe de mostrarse correctanmente 1`] = `
<Fragment>
  <ul
    className="list-group list-group-flush"
  >
    <TodoListItem
      handleDelete={[MockFunction]}
      handleToggle={[MockFunction]}
      i={0}
      key="1"
      todo={
        Object {
          "desc": "Aprender React",
          "done": false,
          "id": 1,
        }
      }
    />
    <TodoListItem
      handleDelete={[MockFunction]}
      handleToggle={[MockFunction]}
      i={1}
      key="2"
      todo={
        Object {
          "desc": "Aprender Next",
          "done": false,
          "id": 2,
        }
      }
    />
  </ul>
</Fragment>
`;
